### DISCUSSION TITLE
Error running Invoke-PSRule

### DISCUSSION THREAD

**Discussion by paulmcclbg** [microsoft/PSRule#1722] [Q&A]
If I run the following:

**Assert-PSRule** -InputPath ./test-sample-1.json -Module PSRule.Rules.Azure

then the command runs fine and I get rule results outputted to the terminal (8 passed, 2 failed)

As soon as I change to 

**Invoke-PSRule** -InputPath ./test-sample-1.json -Module PSRule.Rules.Azure
I get this error:

_format-default: Cannot process argument because the value of argument "name" is not valid. Change the value of the "name" argument and run the operation again._

Can you please explain ? I want to use Invoke-PSRule due to invokation through Pester.

Note: **Test-PSRuleTarget** also runs fine and returns False.

**Comment by BernieWhite** [comment]
@paulmcclbg Haven't seen that error before. Are you using a PSRule v2.9.0 or another version?

Maybe try `Invoke-PSRule -InputPath ./test-sample-1.json -Module PSRule.Rules.Azure -Format Json`.

What is the content of the JSON file?

If it is an ARM template or compiled Bicep file use `Invoke-PSRule -InputPath ./test-sample-1.json -Module PSRule.Rules.Azure -Format File` as expansion only works with the format of `File`.

**Reply by paulmcclbg** [reply]
@BernieWhite in answer to your questions:

- v2.9.0
- contents were a vnet resource grabbed from an actual environment using the Export command and then saved to a test json file.

**Comment by BernieWhite** [comment]
@paulmcclbg Ok I can reproduce that. It's a bug. It's an issue with the default formatter view. Logged as #1723.

I think this would only be affecting if you are manually running the cmdlet, and in Pester it may not be an issue.

To work around it append a select for the fields you are after, for example:

```powershell
Invoke-PSRule -InputPath ./test-sample-1.json -Module PSRule.Rules.Azure | Select-Object RuleName, Outcome, Recommendation
```

**Reply by paulmcclbg** [reply]
@BernieWhite yes I was running manually and your workaround was successful.

Reason for running manually was experimentation. We have created some custom rules, and was wanting to write some pester tests to check the rules worked as expected. so was trying the relevant commands first before attempting in a script (I've seen the pester tests in the code so will base on those).

As an aside I cannot get the -Type filter on the Rule to work. e.g. Rule 'VNet' -Type 'Microsoft.Network/virtualNetworks' ...

This field in the json is called type , e.g.:

"type": "Microsoft.Network/virtualNetworks"

I've added this binding to the ps-rule.yaml:

binding:
  targetType:
    - type

But the rule is ignored. As soon as I remove the -Type filter the rule is executed.

What am I doing wrong ?

**Reply by BernieWhite** [reply]
@paulmcclbg That should work. However, try the following configuration for binding in `ps-rule.yaml`.

```yaml
binding:
  preferTargetInfo: true
  targetName:
    - ResourceName
    - name
  targetType:
    - ResourceType
    - type
```

If this does not work. Try running the command with `-Verbose -Debug` which may reveal additional information that will help us troubleshoot the issue.

**Reply by BernieWhite** [reply]
This issue is fixed in PowerShell v7.4.2. Affected versions of PowerShell include v7.4.0 - v7.4.1.
